OpenRA.Game/FileFormats/Png.cs

```diff
@@ -0,0 +1,211 @@
+#region Copyright & License Information
+/*
+ * Copyright 2007-2018 The OpenRA Developers (see AUTHORS)
+ * This file is part of OpenRA, which is free software. It is made
+ * available to you under the terms of the GNU General Public License
+ * as published by the Free Software Foundation, either version 3 of
+ * the License, or (at your option) any later version. For more
+ * information, see COPYING.
+ */
+#endregion
+
+using System;
+using System.Collections.Generic;
+using System.Drawing;
+using System.Drawing.Imaging;
+using System.IO;
+using System.IO.Compression;
+using System.Linq;
+using System.Net;
+using System.Text;
+
+namespace OpenRA.FileFormats
+{
+	public class Png
+	{
+		public int Width { get; set; }
+		public int Height { get; set; }
+		public PixelFormat PixelFormat { get; set; }
+		public Color[] Palette { get; set; }
+		public byte[] Data { get; set; }
+		public Dictionary<string, string> Meta = new Dictionary<string, string>();
+
+		public Png(Stream s)
+		{
+			if (!Verify(s))
+				throw new InvalidDataException("PNG Signature is bogus");
+
+			s.Position += 8;
+			var headerParsed = false;
+
+			using (var br = new BinaryReader(s))
+			{
+				var data = new List<byte>();
+
+				for (;;)
+				{
+					var length = IPAddress.NetworkToHostOrder(br.ReadInt32());
+					var type = Encoding.UTF8.GetString(br.ReadBytes(4));
+					var content = br.ReadBytes(length);
+					/*var crc = */br.ReadInt32();
+
+					if (!headerParsed && type != "IHDR")
+						throw new InvalidDataException("Invalid PNG file - header does not appear first.");
+
+					using (var ms = new MemoryStream(content))
+					using (var cr = new BinaryReader(ms))
+						switch (type)
+						{
+							case "IHDR":
+								{
```

Comment:
Style nit: to improve readability we prefer to indent switch cases like so:
```csharp
switch (type)
{
	case "IHDR":
	{
		if (headerParsed)
			/* etc */
	}
}
```
(minus one indentation level on the braces)